stl/inc/atomic

```diff
@@ -480,13 +490,37 @@ struct _Atomic_storage<_Ty, 1> { // lock-free using 1-byte intrinsics
 
     bool compare_exchange_strong(_Ty& _Expected, const _Ty _Desired,
         const memory_order _Order = memory_order_seq_cst) noexcept { // CAS with given memory order
-        const char _Expected_bytes = _Atomic_reinterpret_as<char>(_Expected); // read before atomic operation
+        char _Expected_bytes = _Atomic_reinterpret_as<char>(_Expected); // read before atomic operation
         char _Prev_bytes;
-        _ATOMIC_CHOOSE_INTRINSIC(_Order, _Prev_bytes, _InterlockedCompareExchange8, _Atomic_address_as<char>(_Storage),
-            _Atomic_reinterpret_as<char>(_Desired), _Expected_bytes);
-        if (_Prev_bytes == _Expected_bytes) {
-            return true;
+
+#if _HAS_CXX20
+        if constexpr (_STD has_unique_object_representations_v<_Ty>) {
+#endif
+            _ATOMIC_CHOOSE_INTRINSIC(_Order, _Prev_bytes, _InterlockedCompareExchange8,
+                _Atomic_address_as<char>(_Storage), _Atomic_reinterpret_as<char>(_Desired), _Expected_bytes);
+            if (_Prev_bytes == _Expected_bytes) {
+                return true;
+            }
+#if _HAS_CXX20
+        } else {
+            _Ty _Mask;
+            _CSTD memset(&_Mask, 0xff, sizeof(_Ty));

```

Comment:
I think `&_Mask` needs to be `_STD addressof(_Mask)` (which appears on the line below). Here, `_Ty` is `memset`-able, but it could still overload `operator&` without affecting its triviality. (This may indicate that we don't have test coverage for `atomic<Evil>` where `Evil` overloads `operator&`.)

Occurs below.

Reply:
Aha, we do have test coverage (indirectly), and the tests are failing:

```
In file included from C:\agent\_work\1\s\tests\std\tests\VSO_0000000_instantiate_algorithms_32_difference_type_2\test.cpp:12:
In file included from C:/agent/_work/1/s/tests/std/include\instantiate_algorithms_op_deref.hpp:8:
In file included from C:/agent/_work/1/s/tests/std/include\instantiate_algorithms.hpp:10:
In file included from D:\build\x64\out\inc\functional:17:
In file included from D:\build\x64\out\inc\memory:18:
D:\build\x64\out\inc\atomic(507,26): error: overload resolution selected deleted operator '&'
            _CSTD memset(&_Mask, 0xff, sizeof(_Ty));
                         ^~~~~~
D:\build\x64\out\inc\atomic(1745,22): note: in instantiation of member function 'std::_Atomic_storage<Meow::BasicRanIt<false>, 1>::compare_exchange_strong' requested here
        return this->compare_exchange_strong(_Expected, _Desired);
                     ^
D:\build\x64\out\inc\execution(363,25): note: in instantiation of member function 'std::atomic<Meow::BasicRanIt<false> >::compare_exchange_weak' requested here
        while (!_Result.compare_exchange_weak(_Expected, _Local_result) && _Expected > _Local_result) { // keep trying
                        ^
D:\build\x64\out\inc\execution(1307,18): note: in instantiation of member function 'std::_Parallel_choose_min_result<Meow::BasicRanIt<false> >::_Imbue' requested here
        _Results._Imbue(_Key._Chunk_number, _This_find);
                 ^
D:\build\x64\out\inc\execution(202,23): note: in instantiation of member function 'std::_Static_partitioned_find2<Meow::BasicRanIt<false>, (lambda at D:\build\x64\out\inc\execution:1348:13)>::_Process_chunk' requested here
    while (_Operation._Process_chunk() == _Cancellation_status::_Running) { // process while there are chunks remaining
                      ^
D:\build\x64\out\inc\execution(213,5): note: in instantiation of function template specialization 'std::_Run_available_chunked_work<std::_Static_partitioned_find2<Meow::BasicRanIt<false>, (lambda at D:\build\x64\out\inc\execution:1348:13)> >' requested here
    _Run_available_chunked_work(_Operation);
    ^
D:\build\x64\out\inc\execution(1328,17): note: (skipping 2 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
                _Run_chunked_parallel_work(_Hw_threads, _Operation);
                ^
C:/agent/_work/1/s/tests/std/include\instantiate_algorithms.hpp(449,25): note: in instantiation of function template specialization 'std::find<const std::execution::sequenced_policy &, Meow::BasicRanIt<true>, Meow::Evil, 0>' requested here
            (void) std::find(std::forward<ExecutionPolicy>(exec), fwd1, fwd1, VAL);
                        ^
C:/agent/_work/1/s/tests/std/include\instantiate_algorithms.hpp(503,13): note: in instantiation of function template specialization 'std_testing::algorithms_tester<Meow::Evil, Meow::BasicInIt<true>, Meow::BasicFwdIt<true>, Meow::BasicBidIt<true>, Meow::BasicRanIt<true>, Meow::BasicOutIt<true>, std_testing::dummy_functors::Predicate<Meow::Evil>, std_testing::dummy_functors::BiPredicate<Meow::Evil>, std_testing::dummy_functors::UnaryOperation<Meow::Evil>, std_testing::dummy_functors::BinaryOperation<Meow::Evil>, std_testing::dummy_functors::Generator<Meow::Evil>, std_testing::dummy_functors::Function<Meow::Evil>, std_testing::dummy_functors::Rng, std_testing::dummy_functors::Urng>::test_exec_fwd1<Meow::BasicRanIt<true>, const std::execution::sequenced_policy &>' requested here
            test_exec_fwd1(std::execution::seq, fwd1);
            ^
C:/agent/_work/1/s/tests/std/include\instantiate_algorithms.hpp(741,13): note: in instantiation of function template specialization 'std_testing::algorithms_tester<Meow::Evil, Meow::BasicInIt<true>, Meow::BasicFwdIt<true>, Meow::BasicBidIt<true>, Meow::BasicRanIt<true>, Meow::BasicOutIt<true>, std_testing::dummy_functors::Predicate<Meow::Evil>, std_testing::dummy_functors::BiPredicate<Meow::Evil>, std_testing::dummy_functors::UnaryOperation<Meow::Evil>, std_testing::dummy_functors::BinaryOperation<Meow::Evil>, std_testing::dummy_functors::Generator<Meow::Evil>, std_testing::dummy_functors::Function<Meow::Evil>, std_testing::dummy_functors::Rng, std_testing::dummy_functors::Urng>::test_fwd1<Meow::BasicRanIt<true> >' requested here
            test_fwd1(RANIT);
            ^
C:/agent/_work/1/s/tests/std/include\instantiate_algorithms.hpp(829,25): note: in instantiation of member function 'std_testing::algorithms_tester<Meow::Evil, Meow::BasicInIt<true>, Meow::BasicFwdIt<true>, Meow::BasicBidIt<true>, Meow::BasicRanIt<true>, Meow::BasicOutIt<true>, std_testing::dummy_functors::Predicate<Meow::Evil>, std_testing::dummy_functors::BiPredicate<Meow::Evil>, std_testing::dummy_functors::UnaryOperation<Meow::Evil>, std_testing::dummy_functors::BinaryOperation<Meow::Evil>, std_testing::dummy_functors::Generator<Meow::Evil>, std_testing::dummy_functors::Function<Meow::Evil>, std_testing::dummy_functors::Rng, std_testing::dummy_functors::Urng>::test' requested here
        test_algorithms.test();
                        ^
C:/agent/_work/1/s/tests/std/include\instantiate_algorithms_op_deref.hpp(309,18): note: in instantiation of function template specialization 'std_testing::instantiate_std_algorithms_with<Meow::Evil, Meow::BasicInIt<true>, Meow::BasicFwdIt<true>, Meow::BasicBidIt<true>, Meow::BasicRanIt<true>, Meow::BasicOutIt<true>, std_testing::dummy_functors::Predicate<Meow::Evil>, std_testing::dummy_functors::BiPredicate<Meow::Evil>, std_testing::dummy_functors::UnaryOperation<Meow::Evil>, std_testing::dummy_functors::BinaryOperation<Meow::Evil>, std_testing::dummy_functors::Generator<Meow::Evil>, std_testing::dummy_functors::Function<Meow::Evil>, std_testing::dummy_functors::Rng, std_testing::dummy_functors::Urng>' requested here
    std_testing::instantiate_std_algorithms_with<Meow::Evil>(Meow::BasicInIt<true>{11, 22}, Meow::BasicFwdIt<true>{},
                 ^
C:/agent/_work/1/s/tests/std/include\instantiate_algorithms_op_deref.hpp(303,10): note: candidate function [with T = Meow::BasicRanIt<false>] has been explicitly deleted
    void operator&(const T&) = delete;
         ^
```